README.md
## About

Read metrics from Epever Tracer via RS485 and output to InfluxDB.

<img src="https://github.com/dernasherbrezon/epever-tracer-influxdb/raw/master/docs/screen1.png">

## Install

1. Make sure Epever Tracer connected via RS485. If you're using [standard](https://www.amazon.com/gp/product/B016RU8JUY/ref=oh_aui_search_detailpage?ie=UTF8&psc=1) USB->RS485 cable, then you have to install custom drivers:

```bash
cd xr_usb_serial_common
sudo ./build_and_install.sh
```

2. Configure external apt repo:

```bash
sudo add-apt-repository ppa:rodionovamp/epever-tracer-influxdb
sudo apt-get update
```

3. Install application:

```bash
sudo apt-get install epever-tracer-influxdb
```

4. Configure:

```bash
sudo vi /etc/epeverTracer.conf
```

## Build

1. Install dependencies

```bash
sudo apt-get install libmodbus-dev check cmake
```

2. Build

```bash
mkdir build
cd build
cmake ..
make
```

debian/changelog
epever-tracer-influxdb (1.0.3-3) focal; urgency=medium

  * bump build

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Thu, 13 Aug 2020 09:12:21 +0200

epever-tracer-influxdb (1.0.3-2) disco; urgency=medium

  * bump build

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Thu, 13 Aug 2020 09:12:21 +0200

epever-tracer-influxdb (1.0.3-1) disco; urgency=medium

  * fix cron issue

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Thu, 13 Aug 2020 09:12:21 +0200

epever-tracer-influxdb (1.0.2-5) disco; urgency=medium

  * bump armhf build

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Tue, 24 Mar 2020 14:17:21 +0100

epever-tracer-influxdb (1.0.2-4) disco; urgency=medium

  * fix test

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Tue, 24 Mar 2020 13:17:21 +0100

epever-tracer-influxdb (1.0.1-3) disco; urgency=medium

  * fix test

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Tue, 24 Mar 2020 12:17:21 +0100

epever-tracer-influxdb (1.0.0-2) disco; urgency=medium

  * fix packaging

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Mon, 23 Mar 2020 12:17:21 +0100

epever-tracer-influxdb (1.0.0-1) disco; urgency=medium

  * Initial release. (Closes: #1)

 -- Andrey Rodionov <dernasherbrezon@gmail.com>  Mon, 23 Mar 2020 10:17:21 +0100


test/MockEpeverTracer.c
#include <modbus/modbus-rtu.h>
#include <modbus/modbus.h>

#define modbus_new_rtu modbus_new_rtu_mocked
#define modbus_set_slave modbus_set_slave_mocked
#define modbus_free modbus_free_mocked
#define modbus_set_debug modbus_set_debug_mocked
#define modbus_set_error_recovery modbus_set_error_recovery_mocked
#define modbus_set_response_timeout modbus_set_response_timeout_mocked
#define modbus_connect modbus_connect_mocked
#define modbus_close modbus_close_mocked
#define gethostname gethostname_mocked
#define malloc malloc_mocked
#define modbus_read_input_registers modbus_read_input_registers_mocked

modbus_t* modbus_new_rtu_mocked(const char *device, int baud, char parity,
                                    int data_bit, int stop_bit);
int modbus_set_slave_mocked(modbus_t* ctx, int slave);
void modbus_free_mocked(modbus_t *ctx);
int modbus_set_debug_mocked(modbus_t *ctx, int flag);
int modbus_set_error_recovery_mocked(modbus_t *ctx, modbus_error_recovery_mode error_recovery);
int modbus_set_response_timeout_mocked(modbus_t *ctx, uint32_t to_sec, uint32_t to_usec);
int modbus_connect_mocked(modbus_t *ctx);
void modbus_close_mocked(modbus_t *ctx);
int	gethostname_mocked(char *, size_t);
void *malloc_mocked(size_t size);
int modbus_read_input_registers_mocked(modbus_t *ctx, int addr, int nb, uint16_t *dest);

#include "../src/EpeverTracer.c"

#undef modbus_new_rtu
#undef modbus_set_slave
#undef modbus_free
#undef modbus_set_debug
#undef modbus_set_error_recovery
#undef modbus_set_response_timeout
#undef modbus_connect
#undef modbus_close
#undef gethostname
#undef malloc

struct _modbus {
	int dummy;
};

typedef struct {
	int addr;
	uint16_t value;
} mock_register;

typedef struct {
	int failmodbus_new_rtu;
	int failmodbus_set_slave;
	int failmodbus_set_debug;
	int failmodbus_set_error_recovery;
	int failmodbus_set_response_timeout;
	int failmodbus_connect;
	int failgethostname;
	char *hostname;
	int failmalloc;
	mock_register *registers;
	size_t registers_size;
} mock_status;

mock_status mock;

modbus_t* modbus_new_rtu_mocked(const char *device, int baud, char parity, int data_bit, int stop_bit) {
	if (mock.failmodbus_new_rtu) {
		return NULL;
	}
	return (modbus_t*) malloc(sizeof(modbus_t));
}

int modbus_set_slave_mocked(modbus_t *ctx, int slave) {
	if (mock.failmodbus_set_slave) {
		return -1;
	}
	return 0;
}

int modbus_set_debug_mocked(modbus_t *ctx, int flag) {
	if (mock.failmodbus_set_debug) {
		return -1;
	}
	return 0;
}

int modbus_set_error_recovery_mocked(modbus_t *ctx, modbus_error_recovery_mode error_recovery) {
	if (mock.failmodbus_set_error_recovery) {
		return -1;
	}
	return 0;
}

int modbus_set_response_timeout_mocked(modbus_t *ctx, uint32_t to_sec, uint32_t to_usec) {
	if (mock.failmodbus_set_response_timeout) {
		return -1;
	}
	return 0;
}

int modbus_connect_mocked(modbus_t *ctx) {
	if(mock.failmodbus_connect) {
		return -1;
	}
	return 0;
}

int	gethostname_mocked(char *hostname, size_t maxSize) {
	if( mock.failgethostname ) {
		return -1;
	}
	strcpy(hostname, mock.hostname);
	return 0;
}

void* malloc_mocked(size_t size) {
	if (mock.failmalloc) {
		return NULL;
	}
	return malloc(size);
}

int modbus_read_input_registers_mocked(modbus_t *ctx, int addr, int nb, uint16_t *dest) {
	for (int i = 0; i < mock.registers_size; i++) {
		if (mock.registers[i].addr == addr) {
			*dest = mock.registers[i].value;
			return 0;
		}
	}
	return -1;
}

void modbus_free_mocked(modbus_t *ctx) {
	free(ctx);
}

void modbus_close_mocked(modbus_t *ctx) {
	//do nothing
}

void epever_tracer_mock_reset() {
	mock.failmodbus_new_rtu = 0;
	mock.failmodbus_set_slave = 0;
	mock.failmodbus_set_debug = 0;
	mock.failmodbus_set_error_recovery = 0;
	mock.failmodbus_set_response_timeout = 0;
	mock.failmodbus_connect = 0;
	mock.failgethostname = 0;
	mock.hostname = "mockhost";
	mock.failmalloc = 0;
	if (mock.registers != NULL) {
		free(mock.registers);
	}
	mock.registers_size = 0;
}


test/test_EpeverTracer.c
#include <stdlib.h>
#include <check.h>
#include "MockEpeverTracer.c"

START_TEST (test_unableToCreateModBus) {
	mock.failmodbus_new_rtu = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, stdout), EXIT_FAILURE);
}
END_TEST

START_TEST (test_failToSetSlave) {
	mock.failmodbus_set_slave = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, stdout), EXIT_FAILURE);
}
END_TEST

START_TEST (test_failToSetDebug) {
	mock.failmodbus_set_debug = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 1, stdout), EXIT_FAILURE);
}
END_TEST

START_TEST (test_failToSetErrorRecovery) {
	mock.failmodbus_set_error_recovery = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, stdout), EXIT_FAILURE);
}
END_TEST

START_TEST (test_failToSetTimeout) {
	mock.failmodbus_set_response_timeout = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, stdout), EXIT_FAILURE);
}
END_TEST

START_TEST (test_failToConnect) {
	mock.failmodbus_connect = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, stdout), EXIT_FAILURE);
}
END_TEST

START_TEST (test_failGetHostname) {
	mock.failgethostname = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, stdout), EXIT_FAILURE);
}
END_TEST

START_TEST (test_failMalloc) {
	mock.failmalloc = 1;
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, stdout), EXIT_FAILURE);
}
END_TEST

void assertFiles(char *expected, char *actual) {
	FILE *expectedPtr = fopen(expected, "r");
	ck_assert_ptr_ne(expectedPtr, NULL);
	FILE *actualPtr = fopen(actual, "r");
	ck_assert_ptr_ne(actualPtr, NULL);
	int expectedChar;
	int actualChar;
	do {
		expectedChar = fgetc(expectedPtr);
		actualChar = fgetc(actualPtr);
		ck_assert_uint_eq(expectedChar, actualChar);
	} while (expectedChar != EOF && actualChar != EOF);
	fclose(expectedPtr);
	fclose(actualPtr);
}

START_TEST (test_success) {
	char *actual = "test_success.txt";
	FILE *output = fopen(actual, "w");
	mock.registers_size = 5;
	mock.registers = (mock_register*) malloc(mock.registers_size * sizeof(mock_register));
	mock.registers[0] = (mock_register ) { 0x3104, 1240 };
	mock.registers[1] = (mock_register ) { 0x3101, 110 };
	mock.registers[2] = (mock_register ) { 0x3002, 0x93E0 };
	mock.registers[3] = (mock_register ) { 0x3003, 0x0004 };
	mock.registers[4] = (mock_register ) { 0x3200, 0x0001 };
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, output), EXIT_SUCCESS);
	fflush(output);
	fclose(output);
	assertFiles("expected.txt", actual);
	remove(actual);
}
END_TEST

int getTotalRegisters() {
	int result = 0;
	int size = sizeof(registers) / sizeof(registers[0]);
	for (int i = 0; i < size; i++) {
		result += registers[i].size;
	}
	return result;
}

START_TEST (test_allMetrics) {
	char *actual = "test_allMetrics.txt";
	FILE *output = fopen(actual, "w");
	mock.registers_size = getTotalRegisters();
	mock.registers = (mock_register*) malloc(mock.registers_size * sizeof(mock_register));
	for (int sourceIndex = 0, destIndex = 0; destIndex < mock.registers_size; sourceIndex++) {
		for (int j = 0; j < registers[sourceIndex].size; j++) {
			mock.registers[destIndex] = (mock_register ) { registers[sourceIndex].addr + j, 0 };
			destIndex++;
		}
	}
	ck_assert_int_eq(epever_tracer_process("/dev/ttyXRUSB0", 5, 0, output), EXIT_SUCCESS);
	fflush(output);
	fclose(output);
	assertFiles("expectedallMetrics.txt", actual);
	remove(actual);
}
END_TEST

void teardown() {
	epever_tracer_mock_reset();
}

void setup() {
	//do nothing
}

Suite * common_suite(void) {
	Suite *s;
	TCase *tc_core;

	s = suite_create("common");

	/* Core test case */
	tc_core = tcase_create("Core");

	tcase_add_test(tc_core, test_unableToCreateModBus);
	tcase_add_test(tc_core, test_failToSetSlave);
	tcase_add_test(tc_core, test_failToSetDebug);
	tcase_add_test(tc_core, test_failToSetErrorRecovery);
	tcase_add_test(tc_core, test_failToSetTimeout);
	tcase_add_test(tc_core, test_failToConnect);
	tcase_add_test(tc_core, test_failGetHostname);
	tcase_add_test(tc_core, test_failMalloc);
	tcase_add_test(tc_core, test_success);
	tcase_add_test(tc_core, test_allMetrics);

	tcase_add_checked_fixture(tc_core, setup, teardown);
	suite_add_tcase(s, tc_core);

	return s;
}

int main(void) {
	int number_failed;
	Suite *s;
	SRunner *sr;

	s = common_suite();
	sr = srunner_create(s);

	srunner_set_fork_status(sr, CK_NOFORK);
	srunner_run_all(sr, CK_NORMAL);
	number_failed = srunner_ntests_failed(sr);
	srunner_free(sr);
	return (number_failed == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
}




src/main.c
#include "EpeverTracer.h"
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <getopt.h>

extern int errno;
extern char *optarg;
extern int opterr;

int main(int argc, char *argv[]) {
	int opt;
	opterr = 0;
	char *device = NULL;
	int verbose = 0;
	uint32_t timeoutSecond = 5;
	while ((opt = getopt(argc, argv, "d:t:vh")) != EOF) {
		switch (opt) {
		case 'd':
			device = optarg;
			break;
		case 'v':
			verbose++;
			break;
		case 't':
			timeoutSecond = (uint32_t) strtoul(optarg, NULL, 10);
			break;
		case 'h':
		default:
			fprintf(stderr, "%s [-d device] [-v] [-t timeoutSeconds(optional)] [-h]\n", argv[0]);
			return EXIT_FAILURE;
		}
	}
	if (device == NULL) {
		fprintf(stderr, "device is not specified\n");
		errno = EINVAL;
		return EXIT_FAILURE;
	}
	return epever_tracer_process(device, timeoutSecond, verbose, stdout);
}


src/EpeverTracer.c
#define _GNU_SOURCE

#include "EpeverTracer.h"
#include <modbus/modbus-rtu.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>

typedef struct {
	char *category;
	char *parameter;
	int addr;
	int size;
	int toFloat;
} epever_register_t;

epever_register_t registers[] = {
		{"solar", "voltage", 0x3100, 1, 1},
		{"solar", "current", 0x3101, 1, 1},
		{"solar", "power", 0x3102, 2, 1},
		{"battery", "voltage", 0x3104, 1, 1},
		{"battery", "current", 0x3105, 1, 1},
		{"battery", "power", 0x3106, 2, 1},
		{"battery", "temperature", 0x3110, 1, 1},
		{"battery", "remoteTemperature", 0x311B, 1, 1},
		{"battery", "realRatedVoltage", 0x311D, 1, 1},
		{"battery", "soc", 0x311A, 1, 1},
		{"load", "voltage", 0x310C, 1, 1},
		{"load", "current", 0x310D, 1, 1},
		{"load", "power", 0x310E, 2, 1},
		{"device", "temperature", 0x3112, 1, 1},
		{"ratedSolar", "voltage", 0x3000, 1, 1},
		{"ratedSolar", "current", 0x3001, 1, 1},
		{"ratedSolar", "power", 0x3002, 2, 1},
		{"ratedBattery", "voltage", 0x3004, 1, 1},
		{"ratedBattery", "current", 0x3005, 1, 1},
		{"ratedBattery", "power", 0x3006, 2, 1},
		{"ratedLoad", "current", 0x300E, 1, 1},
		{"battery", "status", 0x3200, 1, 0},
		{"device", "chargingMode", 0x3008, 1, 0},
		{"device", "status", 0x3201, 1, 0}
};

int epever_tracer_process(char *device, uint32_t timeoutSecond, int verbose, FILE *output) {
	// values are defined by the protocol
	modbus_t *modbus = modbus_new_rtu(device, 115200, 'N', 8, 1);
	if (modbus == NULL) {
		fprintf(stderr, "Could not connect to MODBUS: %s\n", modbus_strerror(errno));
		return EXIT_FAILURE;
	}
	int rc = modbus_set_slave(modbus, 1);
	if (rc == -1) {
		fprintf(stderr, "unable to setup slave: %s\n", modbus_strerror(errno));
		modbus_free(modbus);
		return EXIT_FAILURE;
	}
	if (verbose > 0) {
		rc = modbus_set_debug(modbus, TRUE);
		if (rc == -1) {
			fprintf(stderr, "unable to setup debug mode: %s\n", modbus_strerror(errno));
			modbus_free(modbus);
			return EXIT_FAILURE;
		}
	}
	rc = modbus_set_error_recovery(modbus, MODBUS_ERROR_RECOVERY_LINK | MODBUS_ERROR_RECOVERY_PROTOCOL);
	if (rc == -1) {
		fprintf(stderr, "unable to setup error recovery: %s\n", modbus_strerror(errno));
		modbus_free(modbus);
		return EXIT_FAILURE;
	}
	rc = modbus_set_response_timeout(modbus, timeoutSecond, 0);
	if (rc == -1) {
		fprintf(stderr, "unable to setup timeout: %s\n", modbus_strerror(errno));
		modbus_free(modbus);
		return EXIT_FAILURE;
	}
	rc = modbus_connect(modbus);
	if (rc == -1) {
		fprintf(stderr, "connection failed: %s\n", modbus_strerror(errno));
		modbus_free(modbus);
		return EXIT_FAILURE;
	}
	char hostname[1024];
	hostname[1023] = '\0';
	rc = gethostname(hostname, 1023);
	if (rc == -1) {
		fprintf(stderr, "unable to get hostname\n");
		modbus_close(modbus);
		modbus_free(modbus);
		return EXIT_FAILURE;
	}

	int maxSize = 2;
	uint16_t *outputArray = (uint16_t*) malloc(maxSize * sizeof(uint16_t));
	if (outputArray == NULL) {
		modbus_close(modbus);
		modbus_free(modbus);
		return EXIT_FAILURE;
	}

	epever_register_t *ptr = registers;
	const epever_register_t *endPtr = registers + sizeof(registers) / sizeof(registers[0]);
	while (ptr < endPtr) {
		memset(outputArray, 0, maxSize * sizeof(uint16_t));
		for (int i = 0; i < ptr->size; i++) {
			uint16_t dest;
			rc = modbus_read_input_registers(modbus, ptr->addr + i, 0x1, &dest);
			if (rc == -1) {
				fprintf(stderr, "failed to read 0x%x: %s\n", ptr->addr, modbus_strerror(errno));
				break;
			}
			outputArray[i] = dest;
		}
		if (rc == -1) {
			ptr++;
			continue;
		}
		uint32_t fullRegister = (((uint32_t) outputArray[1]) << 16) | (outputArray[0]);
		if (ptr->toFloat == 1) {
			fprintf(output, "%s,host=%s %s=%.2f\n", ptr->category, hostname, ptr->parameter, (float) (fullRegister / 100.0));
		} else {
			fprintf(output, "%s,host=%s %s=%d\n", ptr->category, hostname, ptr->parameter, fullRegister);
		}
		ptr++;
	}
	free(outputArray);
	modbus_close(modbus);
	modbus_free(modbus);
	return EXIT_SUCCESS;
}
